File: go/guessing_game.go
// jalavoy - 03.05.2018
// this program plays a quick guessing game with the user. It picks a number between the globals defined below, and then prompts the user to guess.
package main

import "fmt"
import "time"
import "math/rand"
import "bufio"
import "os"
import "strconv"
import "strings"
import "sort"
import "regexp"

// stats struct
type Stats struct {
	games   int
	guesses int
	scores  []int
}

// globals for randomness and our statistics
var min = 1
var max = 100
var stats = &Stats{
	games:   0,
	guesses: 0,
}

func main() {
	// main game loop -- run until player says otherwise
	for {
		// run game
		doGame(stats)
		// query player to see if they want to keep playing -- if they don't, break the loop
		response := queryPlayer("Do you want to play again? ")
		var regex = regexp.MustCompile(`^(Y|y)`)
		if !regex.MatchString(response) {
			break
		}
	}
	// print statistics
	printStats(stats)
	// exit and return true to OS
	os.Exit(0)
}

func doGame(stats *Stats) bool {
	// predefine our games stats, guess is defined outside of the range of randomness so it wont trigger the loop on it's own
	var (
		guess      int64
		guesscount int
	)
	// generating our randomness target
	target := _random(min, max)
	fmt.Println("\nI'm thinking of a number between", min, "and", max)
	// while the players guess isn't correct
	for guess != target {
		guesscount++
		response := queryPlayer("Your guess? ")
		response = strings.TrimSpace(response)
		guess, _ = strconv.ParseInt(response, 10, 0)
		if guess == target {
			if guesscount == 1 {
				fmt.Println("You got it right in 1 guess!")
			} else {
				fmt.Println("You got it right in", guesscount, "guesses!")
			}
		} else {
			if guess < target {
				fmt.Println("Higher!")
			} else {
				fmt.Println("Lower!")
			}
		}
	}
	// update stats
	stats.recordStats(guesscount)
	return true
}

func queryPlayer(input string) string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(input)
	response, _ := reader.ReadString('\n')
	return response
}

func (stats *Stats) recordStats(guesscount int) bool {
	stats.games++
	stats.guesses += guesscount
	stats.scores = append(stats.scores, guesscount)
	return true
}

func printStats(stats *Stats) bool {
	fmt.Println("\nOverall Results:")
	fmt.Println("Total Games:", stats.games)
	fmt.Println("Total Guesses:", stats.guesses)
	fmt.Println("Average guesses per game:", _getAvg(stats.scores))
	fmt.Println("Best game:", _getBest(stats.scores))
	return true
}

func _random(min, max int) int64 {
	rand.Seed(time.Now().Unix())
	return int64(rand.Intn(max-min) + min)
}

func _getAvg(scores []int) int {
	total := 0
	for _, value := range scores {
		total += value
	}
	avg := total / len(scores)
	return avg
}

func _getBest(scores []int) int {
	sort.Ints(scores)
	return scores[0]
}

File: java/guessing_game.java
// jalavoy - 07.01.2018
// this program plays a quick guessing game with the user. It picks a number between the globals defined below, and then prompts the user to guess.
import java.util.Random;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.regex.*;

public class guessing_game {
    static int min = 1;
    static int max = 100;
    static Stats s = new Stats();
    static Scanner sc = new Scanner(System.in);
    static Pattern p = Pattern.compile("^(Y|y)");

    public static void main(String[] args) {
        while (true) {
            do_game(s);
            String response = query_player("Do you want to play again?");
            if (!p.matcher(response).matches()) {
                break;
            }
        }
        s.print_stats();
        System.exit(0);
    }

    private static void do_game(Stats s) {
        int guess = -1;
        int guesscount = 0;
        int target;
        target = get_random();
        System.out.printf("\nI'm thinking of a number between %d and %d\n", min, max);
        while (guess != target) {
            guesscount++;
            String response = query_player("Your guess?");
            response = response.trim();
            guess = Integer.parseInt(response);
            if (guess == target) {
                if (guesscount == 1) {
                    System.out.println("You got it right in 1 guess!");
                } else {
                    System.out.printf("You got it right in %d guesses!\n", guesscount);
                }
            } else {
                if ( guess < target ) {
                    System.out.println("Higher!");
                } else {
                    System.out.println("Lower!");
                }
            }
        }
        s.record_stats(guesscount);
    }

    private static int get_random() {
        Random rand = new Random();
        int n = rand.nextInt(max) + 1;
        return n;
    }

    private static String query_player(String query) {
        String input;
        System.out.printf("%s ", query);
        input = sc.nextLine();
        return input;
    }    

    private static class Stats {
        int games;
        int guesses;
        ArrayList<Number> scores = new ArrayList<Number>();

        private void record_stats(int guesscount) {
            this.games++;
            this.guesses += guesscount;
            this.scores.add(guesscount);
        }

        private void print_stats() {
            System.out.println("\nOverall Results:");
            System.out.println("Total Games: " + this.games);
            System.out.println("Total Guesses: " + this.guesses);
            System.out.println("Average guesses per game: " + get_average());
            System.out.println("Best game: " + get_best());
        }

        private int get_average() {
            int total = 0;
            for (int i = 0; i < this.scores.size(); i++) {
                total += this.scores.get(i).intValue();
            }
            int average = total / this.scores.size();
            return average;
        }

        private int get_best() {
            return this.scores.get(0).intValue();
        }   
    }
}